src/main/java/org/folio/oaipmh/helpers/storage/SourceRecordStorageHelper.java

```diff
@@ -28,10 +35,38 @@ public String getRecordId(JsonObject entry) {
     return entry.getString(RECORD_ID);
   }
 
+  /**
+   * Returns instance id that is linked to record within 999 field
+   *
+   * @param entry the item returned by source-storage
+   * @return instance id
+   */
+  @Override
+  public String getIdentifierId(final JsonObject entry) {
+    Optional<JsonArray> parsedRecordFields = Optional.ofNullable(entry.getJsonObject(PARSED_RECORD))
+      .map(parsedRecord->parsedRecord.getJsonObject(CONTENT))
+      .map(content-> content.getJsonArray(FIELDS));
+    return parsedRecordFields
+      .flatMap(this::getNineNineNineField)
+      .map(field -> field.getJsonArray(SUBFIELDS))
+      .map(subfields -> subfields.getJsonObject(INSTANCE_ID_POSITION))
+      .map(instanceId -> instanceId.getString(INSTANCE_ID))
+      //if 999 field is missed then it means that record hasn't parsedRecord metadata and will be filtered from result response later
+      .orElse("");
+  }
+
+  private Optional<JsonObject> getNineNineNineField(JsonArray jsonArray){
```

Comment:
Can you name this method have a more meaningful name?

Reply:
Yes, sure